File: src/main/java/org/simbrain/network/neuron_update_rules/AdditiveRule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.neuron_update_rules;

import org.simbrain.network.core.Network.TimeType;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.NeuronUpdateRule;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.neuron_update_rules.interfaces.NoisyUpdateRule;
import org.simbrain.network.util.ScalarDataHolder;
import org.simbrain.util.math.ProbDistributions.UniformDistribution;
import org.simbrain.util.math.ProbabilityDistribution;

/**
 * <b>AdditiveNeuron</b> See Haykin (2002), section 14.5. Used with continuous
 * Hopfield networks.
 */
public class AdditiveRule extends NeuronUpdateRule implements NoisyUpdateRule {

    //TODO: May need clipping and bounds.

    /**
     * Lambda.
     */
    private double lambda = 1.4;

    /**
     * Resistance.
     */
    private double resistance = 1;

    /**
     * Noise generator.
     */
    private ProbabilityDistribution noiseGenerator = new UniformDistribution();

    /**
     * For adding noise to the neuron.
     */
    private boolean addNoise = false;

    @Override
    public TimeType getTimeType() {
        return TimeType.CONTINUOUS;
    }

    @Override
    public AdditiveRule deepCopy() {
        AdditiveRule an = new AdditiveRule();
        an.setLambda(getLambda());
        an.setResistance(getResistance());
        an.setAddNoise(getAddNoise());
        an.noiseGenerator = noiseGenerator.deepCopy();
        return an;
    }

    @Override
    public void apply(Neuron neuron, ScalarDataHolder data) {

        // Update buffer of additive neuron using Euler's method.
        double wtdSum = 0;
        if (neuron.getFanIn().size() > 0) {
            for (int j = 0; j < neuron.getFanIn().size(); j++) {
                Synapse w = neuron.getFanIn().get(j);
                Neuron source = w.getSource();
                wtdSum += (w.getStrength() * g(source.getActivation()));
            }
        }

        double val = neuron.getActivation() + neuron.getNetwork().getTimeStep() * (-neuron.getActivation() / resistance + wtdSum);

        if (addNoise) {
            val += noiseGenerator.nextDouble();
        }

        neuron.setActivation(val);
        neuron.addInputValue(0);
    }

    /**
     * Implements a Hopfield type sigmoidal function.
     *
     * @param x input to function
     * @return output of function
     */
    private double g(final double x) {
        return 2 / Math.PI * Math.atan((Math.PI * lambda * x) / 2);
    }

    public double getLambda() {
        return lambda;
    }

    public void setLambda(final double lambda) {
        this.lambda = lambda;
    }

    public double getResistance() {
        return resistance;
    }

    public void setResistance(final double resistance) {
        this.resistance = resistance;
    }

    @Override
    public ProbabilityDistribution getNoiseGenerator() {
        return noiseGenerator;
    }

    @Override
    public void setNoiseGenerator(final ProbabilityDistribution noise) {
        this.noiseGenerator = noise;
    }

    public boolean getAddNoise() {
        return addNoise;
    }

    public void setAddNoise(final boolean addNoise) {
        this.addNoise = addNoise;
    }

    @Override
    public String getName() {
        return "Additive (Continuous Hopfield)";
    }

}

File: src/main/kotlin/org/simbrain/network/updaterules/AdditiveRule.kt
package org.simbrain.network.updaterules

import org.simbrain.network.core.Network
import org.simbrain.network.core.Neuron
import org.simbrain.network.updaterules.interfaces.NoisyUpdateRule
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.util.UserParameter
import org.simbrain.util.applyFunction
import org.simbrain.util.propertyeditor.CustomTypeName
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.UniformRealDistribution
import smile.math.matrix.Matrix
import kotlin.math.atan

/**
 * Decay-type neuron used with continuous Hopfield networks.
 * See Haykin (2002), section 14.5 and the original Hopfield PNAS article.
 */
@CustomTypeName("Additive (Continuous Hopfield)")
class AdditiveRule : NeuronUpdateRule<EmptyScalarData, EmptyMatrixData>(), NoisyUpdateRule {

    @UserParameter(
        label = "Lambda",
        description = "A gain parameter that controls the steepness of the sigmoid curve. Higher lambda makes the function more like a step function; lower lambda produces a gentler slope",
        minimumValue = 0.0,
        increment = .1,
        order = 10)
    var lambda: Double = 1.4

    @UserParameter(
        label = "Resistance",
        description = "Controls the rate of activation decay; higher resistance slows decay, lower resistance causes the activation to decay more quickly",
        minimumValue = 0.0,
        increment = .1,
        order = 20)
    var resistance: Double = 1.0

    override var noiseGenerator: ProbabilityDistribution = UniformRealDistribution()

    override var addNoise: Boolean = false

    override val timeType: Network.TimeType
        get() = Network.TimeType.CONTINUOUS

    override fun copy(): AdditiveRule {
        val an = AdditiveRule()
        an.lambda = lambda
        an.resistance = resistance
        an.addNoise = addNoise
        an.noiseGenerator = noiseGenerator.copy()
        return an
    }

    context(Network)
    override fun apply(neuron: Neuron, data: EmptyScalarData) {

        neuron.activation += timeStep * (-neuron.activation / resistance + neuron.input)

        if (addNoise) {
            neuron.activation += noiseGenerator.sampleDouble()
        }

    }

    /**
     * Implements a Hopfield type sigmoidal function.
     */
    override fun synapticInputModifier(input: Double): Double {
        return 2 / Math.PI * atan((Math.PI * lambda * input) / 2)
    }

    override fun synapticInputModifier(input: Matrix): Matrix {
        return input.applyFunction(::synapticInputModifier)
    }

    override val name: String
        get() = "Additive (Continuous Hopfield)"

}